### PR TITLE
Validate the typed data before dereferencing the internal struct.

### PR DESCRIPTION
**Pull request by ioquatix**
See <https://bugs.ruby-lang.org/issues/19853> for more context.

### CODE REVIEW

**Comment by nobu** [comment]
Shouldn't it be guaranteed that the return value of `rb_process_status_wait` is `Process::Status` or `nil`?

**Comment by ioquatix** [comment]
> Shouldn't it be guaranteed that the return value of rb_process_status_wait is Process::Status or nil?

No, the fiber scheduler can return anything, or raise an exception.

**Comment by nobu** [comment]
> > Shouldn't it be guaranteed that the return value of rb_process_status_wait is Process::Status or nil?
> 
> No, the fiber scheduler can return anything, or raise an exception.

And `Process::Status.wait` is expected to return it as-is?

**Comment by ioquatix** [comment]
> And Process::Status.wait is expected to return it as-is?

Yes, because it's the pure interface with no side effects (doesn't set `last_status`, etc).

File: process.c [hunk 1]
```diff
@@ -1222,7 +1222,9 @@ rb_waitpid(rb_pid_t pid, int *st, int flags)
     VALUE status = rb_process_status_wait(pid, flags);
     if (NIL_P(status)) return 0;
 
+    rb_check_typeddata(status, &rb_process_status_type);
     struct rb_process_status *data = RTYPEDDATA_DATA(status);
+
```
Comment (nobu) [inline process.c:1227]:
```suggestion
    struct rb_process_status *data = rb_check_typeddata(status, &rb_process_status_type);
```

File: process.c [hunk 2]
```diff
@@ -4748,6 +4750,8 @@ rb_f_system(int argc, VALUE *argv, VALUE _)
 
     if (pid > 0) {
         VALUE status = rb_process_status_wait(pid, 0);
+
+        rb_check_typeddata(status, &rb_process_status_type);
         struct rb_process_status *data = RTYPEDDATA_DATA(status);
```
Comment (nobu) [inline process.c:4755]: Ditto.

```suggestion
        struct rb_process_status *data = rb_check_typeddata(status, &rb_process_status_type);
```

And `NIL_P` check is not needed here?

File: process.c (hunk 1 above)
Comment (ioquatix) [reply process.c:1227]: Ah great, thanks.

File: process.c (hunk 2 above)
Comment (ioquatix) [reply process.c]: That is a good question. My understanding is that `rb_f_system` always blocks, so `nil` cannot be returned. `nil` is only returned for non-blocking `Process#wait` IIRC.

**Comment by nobu** [comment]
BTW, why are `rb_process_status_wait` and `rb_process_status_waitv` non-static?
They are not exported in the headers at least.

**Comment by ioquatix** [comment]
> BTW, why are rb_process_status_wait and rb_process_status_waitv non-static?
> They are not exported in the headers at least

In `io-event` gem, it's used. I believe it should be in public header, if it's not, I'll fix that in a separate PR.

```
#ifdef HAVE_RB_PROCESS_STATUS_WAIT
#define IO_Event_Selector_process_status_wait(pid) rb_process_status_wait(pid)
#else
VALUE IO_Event_Selector_process_status_wait(rb_pid_t pid);
#endif
```

**Comment by nobu** [comment]
> In `io-event` gem, it's used. I believe it should be in public header, if it's not, I'll fix that in a separate PR.

They are not exported from libruby.so.
And does it use also `rb_process_status_waitv` function, not only `rb_process_status_wait`?

**Comment by ioquatix** [comment]
I don't know about `rb_process_status_waitv` but I'll investigate it.

**Pull request merged**